src/Rector/Form/CreateOrUpdateFormDataRector.php

```diff
@@ -0,0 +1,156 @@
+use PHPStan\Type\ObjectType;
+use Rector\Core\Rector\AbstractRector;
+use Rector\FileSystemRector\ValueObject\AddedFileWithContent;
+use Rector\Nette\NodeFactory\ClassWithPublicPropertiesFactory;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see https://doc.nette.org/en/3.1/form-presenter#toc-mapping-to-classes
+ */
+final class CreateOrUpdateFormDataRector extends AbstractRector
+{
+    public function __construct(
+        private ClassWithPublicPropertiesFactory $classWithPublicPropertiesFactory
+    ) {
+    }
+
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Create or update form data class with all fields of Form', [/* TODO */]);
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $className = $node->name->name;
+        $fullClassName = $this->getName($node);
+        $form = $this->findForm($node);
+        if ($form === null) {
+            return null;
+        }
+
+        $formFields = $this->findFormFields($node, $form);
+
+        $formDataClassName = $className . 'FormData';
+        $class = $this->classWithPublicPropertiesFactory->createNode(
+            $fullClassName . 'FormData',
+            $formFields,
+            '\Nette\Utils\ArrayHash',   // TODO configurable in rule
+            ['\Nette\SmartObject']  // TODO configurable in rule
+        );
+
+        $printedClassContent = "<?php\n\n" . $this->betterStandardPrinter->print($class) . "\n";
+
+        $smartFileInfo = $this->file->getSmartFileInfo();
+        $targetFilePath = $smartFileInfo->getRealPathDirectory() . '/' . $formDataClassName . '.php';
+
+        $addedFileWithContent = new AddedFileWithContent($targetFilePath, $printedClassContent);
+        $this->removedAndAddedFilesCollector->addAddedFile($addedFileWithContent);
+
+        return null;
+    }
+
+
+    /**
+     * @TODO extract to some service
+     */
+    private function findForm(Class_ $node): ?Variable
+    {
+        foreach ($node->getMethods() as $method) {
+            foreach ($method->stmts as $stmt) {
+                if (! $stmt instanceof Expression) {
+                    continue;
+                }
+
+                if (! $stmt->expr instanceof Assign) {
+                    continue;
+                }
+
+                $var = $stmt->expr->var;
+                $expr = $stmt->expr->expr;
+
+                if (! $var instanceof Variable) {
+                    continue;
+                }
+
+                if (! $this->isObjectType($expr, new ObjectType('Nette\Forms\Form'))) {
+                    continue;
+                }
+
+                return $var;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * @TODO extract to some service
+     */
+    private function findFormFields(Class_ $node, Variable $form): array
```

Comment:
Hi @TomasVotruba I've extracted these methods to some service and now I want to write some simple tests for them. Something like:
```
<?php

use Nette\Application\UI\Form;

class FormClass
{
    public function create(): Form
    {
        $form = new Form();
        $form->addText('foo', 'Foo');
        $form->addTextArea('bar', 'Bar');
        $form->addInteger('baz', 'Baz');

        return $form;
    }
}
-----
{
    "list of fields founded"
}
```

Is there a way how to parse some string (above `-----`) from fixture the same way like rector does it? I tried simple PhpParser but it seems there is much more work to do (types were not resolved completely etc). So basically I am looking for some service which can be used for it. Is there something like that in tests available?
```
/** @var $nodes nodes with types and all info */
$nodes = $superTurboParserWithAllVisitors->parse($content);
```

Reply:
Great questoin!

I think this might help \Rector\Testing\TestingParser\TestingParser

Reply:
Yeah, but it parses whole file (creates SmartFileInfo from file path), i don't have file, just some content :(

Reply:
I see. Just dump the file to temp or extend the service with method that accepts string

Reply:
FYI: with tmp file it works great! Thanks!